@here/harp-texture-atlas-generator/src/ImageVectorCoders.ts

```diff
@@ -0,0 +1,260 @@
+        // No need to keep vector data, we may expose bitmap encoder only.
+        if (this.m_pixelModified) {
+            return ImageBitmapEncoderConstructor.clone(this.m_bitmap);
+        }
+        // Recreate using bitmap and vector data, thus image may be still resized looselessly.
+        else {
+            return new ImageVectorEncoder(this.m_vectorData, this.m_bitmap.clone());
+        }
+    }
+
+    resize(width: number, height: number): ImageEncoder {
+        // Pixels has been modified, drop vector data and resize using bitmap encoder.
+        if (this.m_pixelModified) {
+            return ImageBitmapEncoderConstructor.create(this.m_bitmap).resize(width, height);
+        }
+        // Pixels not modified yet re-convert vector data to bitmap using new size.
+        else {
+            const bitmapResized: Jimp.Jimp = this.m_bitmap.clone().resize(width, height);
+            return new ImageVectorEncoder(this.m_vectorData, bitmapResized);
+        }
+    }
+
+    /**
+     * Write image to file asynchronously.
+     * The image can be written to disk in SVG (if not changed) or PNG, JPEG and BMP format
+     * (based on the file extension provided) or if no extension is provided the PNG image
+     * format is used.
+     *
+     * @param filePath file storage path.
+     */
+    write(filePath: string): Promise<any> {
+        if (!this.m_pixelModified &&
+            FileSystem.getImageFormat(filePath) === ImageFormat.SVG) {
+            return FileSystem.writeFile(filePath, this.m_vectorData);
+        }
+        else {
+            return new Promise((resolve, reject) => {
+                if (!checkExportPath(filePath)) {
+                    reject(new Error("Failed to write and image: " + filePath +
+                        " only BMP, PNG, JPG formats are supported"));
+                }
+                this.m_bitmap.write(filePath, (err, image) => {
+                    if (err) {
+                        reject(new Error("Failed to write and image: " + filePath));
+                    }
+                    resolve();
+                });
+            });
+        }
+    }
+
+    /**
+     * Write image to file synchronously.
+     * @see write.
+     *
+     * @param filePath file storage path.
+     */
+    writeSync(filePath: string): void {
+        if (!this.m_pixelModified &&
+            FileSystem.getImageFormat(filePath) === ImageFormat.SVG) {
+            return FileSystem.writeFileSync(filePath, this.m_vectorData);
+        }
+        else {
+            if (!checkExportPath(filePath)) {
+                throw new Error("Failed to write and image: " + filePath +
+                    " only BMP, PNG, JPG formats are supported");
+            }
+            const mime = FileSystem.getImageMimeType(filePath);
+            this.m_bitmap.getBuffer(mime, (err, buffer) => {
+                if (err) {
+                    throw new Error("Could not get buffer to write and image: " + filePath);
```

Comment:
This is not sync!
It will throw exception in callback, with no-one listening to it.

This function cannot be sync! Because of this one particular API (m_bitmap.getBuffer), you cannot possibly create sync version of this code.
Is this really needed?

Reply:
This was a undocumented trick with Jimp library -> callback realized at once. I have removed it because it is no longer necessary, so function will throw Error if the file was modified and it is impossible to save changes synchronously, unmodified vector data may be saved although via FileSystem.writeFileSync.